File: go-algorithm/leetcode/problem129.go
package main

// 129. Sum Root to Leaf Numbers
// https://leetcode.com/problems/sum-root-to-leaf-numbers/

func sumNumbers(root *TreeNode) int {
	ans := 0
	var dfs func(node *TreeNode, num int)
	dfs = func(node *TreeNode, num int) {
		if node == nil {
			ans += num
			return
		}
		num = num*10 + node.Val
		if node.Left == nil && node.Right == nil {
			ans += num
		}
		if node.Left != nil {
			dfs(node.Left, num)
		}
		if node.Right != nil {
			dfs(node.Right, num)
		}
	}
	dfs(root, 0)
	return ans
}

File: java-algo/src/main/java/com/leetcode/Problem129.java
package com.leetcode;

import com.common.TreeNode;

/**
 * 129. Sum Root to Leaf Numbers
 * https://leetcode.com/problems/sum-root-to-leaf-numbers/
 */
public class Problem129 {
    public int sumNumbers(TreeNode root) {
        return dfs(root, 0);
    }

    private int dfs(TreeNode node, int prev) {
        if (null == node)
            return 0;
        int cur = prev * 10 + node.val;
        if (null == node.left && null == node.right)
            return cur;
        return dfs(node.left, cur) + dfs(node.right, cur);
    }
}
